### PR TITLE
Fs2 chunk aggregator

### PR DESCRIPTION
**Pull request by agustafson**
I'm not sure if this does exactly what's needed, as there wasn't a test in place and there wasn't much reference to it in the docs. I've added a test and the new logic looks similar to what the previous impl was trying to do.
This is a fix for #873

### CODE REVIEW

**Comment by rossabaker** [comment]
The build failure is not yours, by the way.  #1107 broke things, and has been masked by #1123's intermittent failings.

**Comment by agustafson** [comment]
Thanks @rossabaker - lemme know what you want me to do with it and I'm happy to get that done

File: server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    val service: HttpService = HttpService.lift { _ => Ok().withBody(body) }
+    ChunkAggregator(service).run(Request()).unsafeValue() must beSome.like {
+      case maybeResponse: MaybeResponse =>
+        val response = maybeResponse.orNotFound
+        response.status must_== Ok
+        responseCheck(response)
+    }
+  }
+
+  def encodeUtf8String(str: String): Seq[Byte] = str.getBytes.toSeq
```
Comment (ChristopherDavenport) [inline server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala]: Would prefer a supporting function to be declared above or inside the spec

_(More of a preference thing generally, so anyone feel free to disagree)_

File: server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        response.body.runLog.unsafeValue() must_=== Some(Vector.empty)
+      }
+    }
+
+    "handle a Pass" in {
+      val service: HttpService = HttpService.lift(_ => Pass.now)
+      ChunkAggregator(service).run(Request()).unsafeValue() must beSome(Pass)
+    }
+
+    "handle a single chunk" in {
```
Comment (ChristopherDavenport) [inline server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala]: Would prefer this to be a property test. As all of the methods exposed are easy to test, and this is a fairly minimal spec.

File: server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      val str = encodeUtf8String("hiya")
+      checkResponse(Stream.emits(str)) { response =>
+        response.contentLength must_=== Some(4L)
+        response.body.runLog.unsafeValue() must_=== Some(str.toVector)
+      }
+    }
+
+    "handle multiple chunks" in {
```
Comment (ChristopherDavenport) [inline server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala]: Also probably a candidate for a property test.

File: server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala (hunk 1 above)
Comment (agustafson) [inline server/src/test/scala/org/http4s/server/middleware/ChunkAggregatorSpec.scala]: thanks @ChristopherDavenport - all of your suggestions are bang on the money. I'll make those changes now.

**Comment by rossabaker** [comment]
I don't know what to call a version like 0.16 that aggregates just the unemitable chunks.  Maybe we take this as is and introduce a flag when fs2 supports that behavior.

Would a size limit also be good?  Or maybe that's a separate middleware.

**Comment by agustafson** [comment]
I did wonder about a size limit but was unsure what the implementation
would look like. I'm guessing it would be a case of re-chunking to the size
limit but what would the content length be set to? So I'd agree that should
be a separate middleware which is simply a rechunker (is "rechunker" a
word?) RechunkingAggregator? Rechunkificator?

**Comment by rossabaker** [comment]
It's kind of similar to fs2's `rechunkN` with a really large N.  And limiting the overall size is handled by `EntityLimiter`, whose existence I'd forgotten.  I think `RechunkEntity` might be a good name for this, and it could even be implemented in terms of `rechunkN`.

**Comment by agustafson** [comment]
Ok, so would you like me to simply create `RechunkEntity`? What should we
do with `ChunkAggregator` in that case?

**Comment by rossabaker** [comment]
Does `ChunkAggregator` do anything that `RechunkEntity` wouldn't do?  I imagine `RechunkEntity` as something like this:

```scala
def apply(n: Int = Int.MaxValue)(service: HttpService): HttpService
```

For `n` larger than the body, it produces a single chunk, just like `ChunkAggregator` -- the default would be equivalent to ChunkAggregator.  In case smaller chunks are desired, pick a smaller `n`.  In either case, it's just passing through to `rechunkN`.  And then `ChunkAggregator` can be deprecated in favor of the more generic `RechunkEntity`.

I'll trust your judgment on this.  I haven't used this middleware yet.  I just found `rechunkN` when we were discussing names, and it looked like a more generic form, so it seemed reasonable to me.  I might be talking nonsense. :)

**Comment by agustafson** [comment]
There's no nonsense there - all makes perfect sense!

Yep, `RechunkEntity` would simply call `rechunkN`. I haven't used the
middleware either and as I'd mentioned I couldn't find any tests or refs in
the docs. Perhaps it's worthwhile asking the community what's they might
use if for? In the meantime `RechunkEntity` is dead simple to implement.

**Comment by wedens** [comment]
If it can de-chunk response by removing `Transfer-Encoding: chunked` and putting `Content-Length`, it can be useful when you have to interact with some tools (or they have to interact with your server, to be more precise) that suck at chunked enconding (I've seen such tools).

**Comment by agustafson** [comment]
Ah perfect thanks for that example. Should we add those requirements to the
feature? Removing chunking headers?

**Comment by rossabaker** [comment]
We're not going to know the size of the stream until we run the body.  So the `rechunkN` isn't going to be that useful.  But we could buffer the request body, and then invoke the service with that buffered request:

```scala
Service.lift { req =>
  for {
    // TODO account for the fact we stream Byte, not ByteVector.  We need an efficient way to
    // concatenate Chunk[Byte].
    fullBody <- req.body.foldMonoid[Byte]
    // TODO removeChunkedTransferEncoding can't just remove the Transfer-Encoding header:
    // other transfer encodings may be applied.  It needs to remove chunked from any header it
    // does find.
    // n.b. withBody takes care of adding the Content-Length
    newReq <- removeChunkedTransferEncoding(req.withBody(fullBody))
    resp <- service(newReq)
  } yield (resp)
}
```

That solves @wedens' use case, and is probably more useful than rechunking into arbitrary size N.  In that case, I think it's still `ChunkAggregator`.

File: server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala [hunk 4]
```diff
@@ -1,38 +1,20 @@
...
-  def apply(service: HttpService): HttpService = service.map {
+  def apply(service: HttpService): HttpService = service.flatMapF[MaybeResponse] {
     case response: Response =>
-      val chunks = compact(response.body)
-      if (chunks.nonEmpty) {
-        val h = response.headers.put(`Content-Length`(chunks.head.length))
-        response.copy(body = emitAll(chunks), headers = h)
+      response.body.runLog.map { bytes =>
+        if (bytes.nonEmpty) {
+          val h = response.headers.put(`Content-Length`(bytes.length.toLong))
```
Comment (bryce-anderson) [inline server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala]: If we're going to force a `Content-Length` header, we should also ensure there isn't a `Transfer-Encoding: chunked` header present.

**Comment by ChristopherDavenport** [comment]
Been around a month here, what is our status on this?

**Comment by agustafson** [comment]
To be honest I haven't had much time to look at it but wasn't quite sure
what was needed anyway.

**Comment by rossabaker** [comment]
I think it can look something like [this comment](https://github.com/http4s/http4s/pull/1125#issuecomment-297742330).  I think the `foldMonoid` doesn't compile, but the idea is to buffer the body.  `removeChunkedTransferEncoding` just needs to remove `chunked` if there are multiple transfer encodings (rare), and remove the header entirely if it's the only encoding.

**Comment by agustafson** [comment]
I've started looking at this again and should hopefully have something done this week

**Comment by agustafson** [comment]
Ok, I've taken the feedback above and updated the PR. Let me know any feedback.

File: server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala [hunk 5]
```diff
@@ -2,19 +2,32 @@ package org.http4s
...
-      response.body.runLog.map { bytes =>
-        if (bytes.nonEmpty) {
-          val h = response.headers.put(`Content-Length`(bytes.length.toLong))
-          response.copy(body = Stream.emits(bytes), headers = h)
-        } else response
+  def apply(service: HttpService): HttpService =
+    service.flatMapF[MaybeResponse] {
+      case Pass => Task.now(Pass)
+      case response: Response =>
+        response.body.runLog.flatMap { fullBody =>
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala]: Might be able to make this more efficient by foldMapping into a `ByteVector`, and creating a `ByteVectorChunk` from that.  That would avoid some copying of bytes.  But I think this is correct.

File: server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala [hunk 6]
```diff
@@ -2,19 +2,32 @@ package org.http4s
...
+          else
+            Task.now(response)
+        }
+    }
+
+  private val removeChunkedTransferEncoding: Response => Response = {
+    _.transformHeaders { headers =>
+      headers.flatMap {
+        case e: `Transfer-Encoding` =>
+          NonEmptyList.fromList(e.values.filterNot(_ == TransferCoding.chunked)).map(`Transfer-Encoding`.apply).toList
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala]: Took me a minute to understand this, but then I saw the property test and was pretty sure you were right, then came back and understood this.

I love property tests!

**Comment by aeons** [comment]
There's one here: https://github.com/http4s/http4s/blob/master/core/src/main/scala/org/http4s/util/chunk.scala#L10

File: server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala (hunk 6 above)
Comment (agustafson) [inline server/src/main/scala/org/http4s/server/middleware/ChunkAggregator.scala]: yeah, it's not the loveliest piece of code, but wasn't sure how to make it any clearer. I'll just add a comment to the method to make it a little clearer.

**Comment by agustafson** [comment]
WTAF? How was my test passing if that wasn't compiling. I'll fix and take a look at the `runLog` stuff.

**Comment by agustafson** [comment]
Ok, I've fixed up the compilation errors and have made some changes to use a `ByteVector` buffer

**Pull request merged**